### DISCUSSION TITLE
Hooks values  `originalDoc` and `previousDoc` are badly typed

### DISCUSSION THREAD

**Discussion by peufo** [payloadcms/payload#1851] [Feature Requests & Ideas]
# Bug Report

The values `originalDoc` and `previousDoc` passed to functions of type `FieldHook`, `CollectionBeforeChangeHook` and `CollectionAfterChangeHook` do not match.

1. There is no distinction between `beforeChange` and `afterChange` field hooks. This makes in the `beforeChange` hook, `previousDoc` is typed as defined when it is not (as mentioned in [the documentation](https://payloadcms.com/docs/hooks/fields#arguments-and-return-values))

2. In most cases, the document has "_id" and "__v" fields instead of simple "id".

## Steps to Reproduce

Create this collection config : 

```typescript
import type {
  CollectionConfig,
  CollectionBeforeChangeHook,
  CollectionAfterChangeHook,
  FieldHook,
} from 'payload/types'

/* Result of "payload generate:types" */
interface IDoc {
  id: string
  name?: string
  createdAt: string
  updatedAt: string
}

interface IDocWithNativeFields extends IDoc {
  _id?: string
  __v?: number
}

function testDocType(labels: string[], doc: IDocWithNativeFields) {
  const correct = doc && doc.id && !doc._id && !doc.__v
  console.log(correct ? '✔️' : '❌', labels.join('\t'))
}

const fieldBeforeChange: FieldHook<IDoc> = ({ originalDoc, previousDoc }) => {
  testDocType(['field', 'beforeChange', 'previousDoc'], previousDoc)
  testDocType(['field', 'beforeChange', 'originalDoc'], originalDoc)
}

const fieldAfterChange: FieldHook<IDoc> = ({ originalDoc, previousDoc }) => {
  testDocType(['field', 'afterChange', 'previousDoc'], previousDoc)
  testDocType(['field', 'afterChange', 'originalDoc'], originalDoc)
}

const collectionBeforeChange: CollectionBeforeChangeHook<IDoc> = ({
  originalDoc,
}) => {
  testDocType(['collection', 'beforeChange', 'originalDoc'], originalDoc)
}
const collectionAfterChange: CollectionAfterChangeHook<IDoc> = ({
  previousDoc,
}) => {
  testDocType(['collection', 'afterChange', 'originalDoc'], previousDoc)
}

export const Doc: CollectionConfig = {
  slug: 'doc',
  hooks: {
    beforeChange: [collectionBeforeChange],
    afterChange: [collectionAfterChange],
  },
  fields: [
    {
      name: 'name',
      type: 'text',
      hooks: {
        beforeChange: [fieldBeforeChange],
        afterChange: [fieldAfterChange],
      },
    },
  ],
}

```

See the console output when you create or update a doc: 

```
❌ collection   beforeChange    originalDoc
❌ field        beforeChange    previousDoc
❌ field        beforeChange    originalDoc
❌ field        afterChange     previousDoc
✔️ field        afterChange     originalDoc
❌ collection   afterChange     originalDoc
```

## Other Details

- payload version: "^1.5.4"

I think this is partly in line with this issue #1563 

Thank you for this ambitious and forward-looking project 👏

**Comment by jmikrut** [comment]
Hey @peufo — thanks for bringing this up! 

Yes, it certainly does overlap with #1563. We have some big TS stuff in the works that I am VERY excited about and this will be resolved alongside that roadmap item as you expected!

In the meantime, I'm going to convert this to a discussion because this is more about improving our type accuracy rather than something explicitly broken, but we can keep the conversation going there.

How's that?
